sycl/source/detail/platform_impl.cpp

```diff
@@ -49,6 +50,109 @@ platform_impl_host::get_devices(info::device_type dev_type) const {
+  const char *str = getenv("SYCL_DEVICE_WHITE_LIST");
+  if (!str)
+    return {};
+
+  std::vector<DevDescT> decDescs;
+  const char devNameStr[] = "DeviceName";
+  const char driverVerStr[] = "DriverVersion";
+  decDescs.emplace_back();
+  while (*str != '\0') {
+    const char **dstPtr = nullptr;
+    int *Size = nullptr;
+
+    // -1 to avoid comparing null terminator
+    // TODO: Change devNameStr so it doesn't contain null terminator
+    if (0 == strncmp(devNameStr, str, sizeof(devNameStr) - 1)) {
+      dstPtr = &decDescs.back().devName;
+      Size = &decDescs.back().devNameSize;
+      str += sizeof(devNameStr);
+    } else if (0 == strncmp(driverVerStr, str, sizeof(driverVerStr) - 1)) {
+      dstPtr = &decDescs.back().minDriverVer;
+      Size = &decDescs.back().minDriverVerSize;
+      str += sizeof(driverVerStr);
+    }
+
+    if ('\0' == *str || ',' == *str || !dstPtr) {
+      // Malformed white list.
+      // TODO: throw an exception here.
+      break;
+    }
+
+    *dstPtr = str;
+
+    // Skip value
+    while (*str != '\0' && *str != ',' && *str != '|')
+      ++str;
+
+    *Size = str - *dstPtr;
+
+    // | means that the is another filter
+    if(*str == '|')
+      decDescs.emplace_back();
+
+    ++str;
+  }
+
+  return decDescs;
+}
+
+static void filterWhiteList(vector_class<RT::PiDevice> &pi_devices) {
+  const std::vector<DevDescT> whiteList(getWhiteListDesc());
+  if (whiteList.empty())
+    return;
+
+  for (int i = 0; i < (int)pi_devices.size();) {
+    RT::PiDevice dev = pi_devices[i];
+    const string_class devName =
+        sycl::detail::get_device_info<string_class, info::device::name>::_(dev);
+
+    const string_class minDriverVer =
+        sycl::detail::get_device_info<string_class,
+                                      info::device::driver_version>::_(dev);
+
+    bool found = false;
+    for (const DevDescT &desc : whiteList) {
+      // At least device name is required filed to consider the filter
+      if (nullptr == desc.devName ||
+          std::string::npos == devName.find(desc.devName, 0, desc.devNameSize))
+        continue;
+
+      // TODO: Probably comparing strings here is not a good idea as, for
+      // example, the driver version "2019.8.9.0" is greater than "2019.8.10.0"
+      // according to alphabetical order rules. Another option is to convert the
+      // driver version string to integers, but it's not 100% reliable solution
+      // also because CL_DRIVER_VERSION string "follows a vendor-specific
+      // format" according to the OpenCL 2.1 specification.
+      // If no version is specified any version is OK
+      if (nullptr != desc.minDriverVer &&
+          (minDriverVer.compare(0, desc.minDriverVerSize, desc.minDriverVer,
+                                desc.minDriverVerSize) > 0))
+        continue;
+
+      found = true;
+      break;
+    }
+    if (found)
+      continue;
+
+    pi_devices[i] = pi_devices.back();
+    pi_devices.pop_back();
```

Comment:
This is going to change order of devices, right? It should not matter for a properly written program, but can be unexpected. It should be easy to change the loop and iterate forward instead (with two indices: insertion point and current device).

Reply:
Sorry, didn't get the idea how to change the loop.
Furthermore I don't think we should try to preserve the order as program relying on this is 200% invalid. The are source of order change already anyway(icd filenames, OCL_ICD_FILENAMES, OCL_ICD_VENDORS).

Reply:
> I don't think we should try to preserve the order as program relying on this is 200% invalid.

*Relying* on the order is invalid, but let's not break it if we can.

The idea is simple: iterate over devices (index i) and have another index where the next "good" device should be written.
```
int insert_idx = 0;
for (int i = 0; i < devices.size(); ++i) {
  if (isgood(devices[i]) {
    devices[insert_idx++] = devices[i];
  }
}
devices.resize(insert_idx);
```

Reply:
Ok, will do.

Reply:
Done.